src/address.c

```diff
@@ -0,0 +1,145 @@
+        // where key_hash = ripemd160(sha256(pubkey))
+        unsigned char redeem_script[HASH160_LEN + 2];
+        redeem_script[0] = OP_0;
+        redeem_script[1] = HASH160_LEN;
+
+        if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), redeem_script + 2, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+
+        // P2SH address = version (e.g. 0x05) + ripemd160(sha256(redeem_script))
+        address[0] = (unsigned char) version & 0xff;
+        if (wally_hash160(redeem_script, sizeof(redeem_script), address + 1, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+    }
+
+    ret = wally_base58_from_bytes(address, sizeof(address), BASE58_FLAG_CHECKSUM, output);
+
+    wally_clear(address, sizeof(address));
+    return ret;
+}
+
+int wally_bip32_key_to_addr_segwit(const struct ext_key *hdkey, const char *addr_family,
+                                   uint32_t flags, char **output)
+{
+    int ret;
+
+    // Witness program bytes, including the version and data push opcode.
+    unsigned char witness_program_bytes[HASH160_LEN + 2];
+    witness_program_bytes[0] = OP_0;
+    witness_program_bytes[1] = HASH160_LEN;
+
+    if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), witness_program_bytes + 2, HASH160_LEN) != WALLY_OK)
+        return WALLY_EINVAL;
+
+    ret = wally_addr_segwit_from_bytes(witness_program_bytes, HASH160_LEN + 2, addr_family, flags, output);
+    return ret;
+}
+
+int wally_address_to_scriptpubkey(const char *addr, uint32_t flags, unsigned char *bytes_out,
+                                  size_t len, size_t *written)
+{
+    int ret;
+
+    uint32_t version;
+    unsigned char bytes_base58_decode[len + 1];
+    size_t written_base58_decode;
+
+    if (written)
+        *written = 0;
+
+    if ((flags & ~0xff))
+        return WALLY_EINVAL;
+
+    if (!(flags & (WALLY_NETWORK_BITCOIN_MAINNET | WALLY_NETWORK_BITCOIN_TESTNET)))
+        return WALLY_EINVAL;
+
+    // Check minimum length of bytes_out:
+    // P2PKH 25 bytes: OP_DUP OP_HASH160 [pub_key_hash] OP_EQUALVERIFY OP_CHECKSIG
+    // P2SH  23 bytes: OP_HASH160 [script_hash] OP_EQUAL
+    if (len < 25)
+        return WALLY_EINVAL;
+
+    if (wally_base58_to_bytes(addr, BASE58_FLAG_CHECKSUM, bytes_base58_decode, sizeof(bytes_base58_decode), &written_base58_decode) != WALLY_OK)
+        return WALLY_EINVAL;
+
+    if (written_base58_decode == 0)
+        return WALLY_EINVAL;
+
+    version = bytes_base58_decode[0];
+
+    switch (version) {
+    case 0: // Bitcoin P2PKH
+        if (flags & WALLY_NETWORK_BITCOIN_TESTNET)
+            return WALLY_EINVAL;
+    case 111: // Bitcoin testnet P2PKH
+        if (version == 111 && flags & WALLY_NETWORK_BITCOIN_MAINNET)
+            return WALLY_EINVAL;
+        *written = HASH160_LEN + 5;
+        bytes_out[0] = OP_DUP;
+        bytes_out[1] = OP_HASH160;
+        bytes_out[2] = HASH160_LEN;
+        memcpy(bytes_out + 3, bytes_base58_decode + 1, *written);
```

Comment:
Should this be copying HASH160_LEN bytes (rather than HASH160_LEN + 5)?

Reply:
Yes, thanks!